File: 441.arranging-coins.cpp
/*
 * @lc app=leetcode id=441 lang=cpp
 *
 * [441] Arranging Coins
 */

#include <bits/stdc++.h>

#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int arrangeCoins(int n) {
    int64_t sum_k = 0;
    for (int k = 0; k <= n; k++) {
      sum_k += (int64_t)k;
      if (sum_k == n) {
        return k;
      } else if (sum_k > n) {
        return k - 1;
      }
    }
    return 0;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.arrangeCoins(0) == 0);
  assert(s.arrangeCoins(1) == 1);
  assert(s.arrangeCoins(6) == 3);
  assert(s.arrangeCoins(5) == 2);
  assert(s.arrangeCoins(8) == 3);
  assert(s.arrangeCoins(1804289383) == 60070);
  return 0;
}

File: 441.arranging-coins.rs
/*
 * @lc app=leetcode id=441 lang=rust
 *
 * [441] Arranging Coins
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn arrange_coins(n: i32) -> i32 {
        let n = n as i64;
        let mut sum: i64 = 0;
        for k in 0i64..=n {
            sum += k;
            if sum == n {
                return k as i32;
            } else if sum > n {
                return (k - 1) as i32;
            }
        }
        0
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::arrange_coins(0), 0);
    assert_eq!(Solution::arrange_coins(1), 1);
    assert_eq!(Solution::arrange_coins(6), 3);
    assert_eq!(Solution::arrange_coins(5), 2);
    assert_eq!(Solution::arrange_coins(8), 3);
    assert_eq!(Solution::arrange_coins(1804289383), 60070);
}
